docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+```
+
+The same field `preprocess` can be added to `melange.emit`.
+
+Here is the list of all extensions supported by Melange:
+
+- `bs.debugger`: insert `debugger` statements
+- `bs.external`: read global values
+- `bs.obj`: create JavaScript object literals
+- `bs.raw`: write raw JavaScript code
+- `bs.re`: insert regular expressions
+
+### Intro to `external`
+
+Most of the system that Melange exposes to communicate with JavaScript is built
+on top of an OCaml language construct called `external`.
+
+`external` is a keyword for declaring a value in OCaml that will [interface with
+C code](https://v2.ocaml.org/manual/intfc.html):
+
+```ocaml
+external my_c_function : int -> string = "someCFunctionName"
+```
+
+It is like a `let` binding, except that the body of an external is a string.
+That string has a specific meaning depending on the context. For native OCaml,
+it usually refers to a C function with that name. For Melange, it refers to the
+functions or values that exist in the runtime JavaScript code, and will be used
+from Melange.
+
+Melange externals are always decorated with certain `[@bs.xxx]` attributes.
+These attributes are listed [above](#list-of-attributes-and-extensions), and
+will be further explain in the next sections.
+
+Once declared, one can use an `external` as a normal value.
+
+Melange external functions are turned into the expected JavaScript values,
+inlined into their callers during compilation, and completely erased afterwards.
+They don’t appear in the JavaScript output, so there are no costs on bundling
+size.
+
+**Note**: it is recommended to use external functions and the `[@bs.xxx]`
+attributes in the interface files as well, as this allows some optimizations
+where the resulting JavaScript values can be directly inlined at the call sites.
+
+#### Special identity external
+
+One external worth mentioning is the following one:
+
+```ocaml
+type foo = string
+type bar = int
+external danger_zone : foo -> bar = "%identity"
+```
+
+This is a final escape hatch which does nothing but convert from the type `foo`
+to `bar`. In the following sections, if you ever fail to write an `external`,
+you can fall back to using this one. But try not to.
+
+Let’s now see all the ways to use JavaScript from Melange.
+
+### Generate raw JavaScript
+
+It is possible to directly write JavaScript code from a Melange file. This is
+unsafe, but it can be useful for prototyping or as a escape hatch.
+
+To do it, we will use the `bs.raw`
+[extension](https://v2.ocaml.org/manual/extensionnodes.html):
+
+```ocaml
+let add = [%bs.raw {|
+  function(a, b) {
+    console.log("hello from raw JavaScript!");
+    return a + b;
+  }
+|}]
+
+let () = Js.log (add 1 2)
+```
+
+The `{||}` strings are called ["quoted
+strings"](https://ocaml.org/manual/lex.html#quoted-string-id). They are similar
+to JavaScript’s template literals, in the sense that they are multi-line, and
+there is no need to escape characters inside the string.
```

Comment:
but they don't support interpolation.

Reply:
But some of the quoted strings can interpolate :P https://reasonml.github.io/en/try?rrjsx=true&reason=DYUwLgBJC8EEQDMD2S4G4BQHSQK4VgG8ArAHwAkRhgkIASMU4gXyA

I am not sure how to tackle this suggestion, I think it'd be better to leave it as is, and not explain any more that these two similarities (multi-line and escaping), but if you have any ideas on how to explain the interpolation limitation while explaining why `{j||j}` does support it, happy to add them.